plugin/pkg/admission/podpreset/admission.go

```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
+			// make sure they are identical or throw an error
+			if !reflect.DeepEqual(found, v) {
+				errs = append(errs, fmt.Errorf("merging env for %s has a conflict on %s: \n%#v\ndoes not match\n%#v\n in container", pp.GetName(), v.Name, v, found))
+			}
 		}
 	}
 
-	return original, nil
+	err := utilerrors.NewAggregate(errs)
+	if err != nil {
+		return nil, err
+	}
+
+	return mergedEnv, err
 }
 
-func mergeEnvFrom(pip *settings.PodPreset, original []api.EnvFromSource) ([]api.EnvFromSource, error) {
-	// if there were no original envfrom just return the pip envfrom
-	if original == nil {
-		return pip.Spec.EnvFrom, nil
+func MergeEnvFrom(envSources []api.EnvFromSource, podPresets []*settings.PodPreset) ([]api.EnvFromSource, error) {
+
+	var mergedEnvFrom []api.EnvFromSource
+
+	mergedEnvFrom = append(mergedEnvFrom, envSources...)
+	for _, pp := range podPresets {
+		mergedEnvFrom = append(mergedEnvFrom, pp.Spec.EnvFrom...)
 	}
 
-	return append(original, pip.Spec.EnvFrom...), nil
+	return mergedEnvFrom, nil
 }
 
-func mergeVolumeMounts(pip *settings.PodPreset, original []api.VolumeMount) ([]api.VolumeMount, error) {
-	// if there were no original volume mount just return the pip volume mount
-	if original == nil {
-		return pip.Spec.VolumeMounts, nil
-	}
+func mergeVolumeMounts(volumeMounts []api.VolumeMount, podPresets []*settings.PodPreset) ([]api.VolumeMount, error) {
 
-	// first key by name
-	orig := map[string]interface{}{}
-	for _, v := range original {
-		orig[v.Name] = v
+	origVolumeMounts := map[string]api.VolumeMount{}
+	volumeMountsByPath := map[string]api.VolumeMount{}
+	for _, v := range volumeMounts {
+		origVolumeMounts[v.Name] = v
+		volumeMountsByPath[v.MountPath] = v
 	}
 
-	// check for conflicts.
-	for _, v := range pip.Spec.VolumeMounts {
-		found, ok := orig[v.Name]
-		if !ok {
-			// if we don't already have it continue
-			continue
-		}
+	mergedVolumeMounts := make([]api.VolumeMount, len(volumeMounts))
+	copy(mergedVolumeMounts, volumeMounts)
+
+	var errs []error
 
-		// make sure they are identical or throw an error
-		if !reflect.DeepEqual(found, v) {
-			return nil, fmt.Errorf("merging volume mounts for %s has a conflict on %s: \n%#v\ndoes not match\n%#v\n in container", pip.GetName(), v.Name, v, found)
+	for _, pp := range podPresets {
+		for _, v := range pp.Spec.VolumeMounts {
+			found, ok := origVolumeMounts[v.Name]
+			if !ok {
+				// if we don't already have it append it and continue
+				origVolumeMounts[v.Name] = v
+				mergedVolumeMounts = append(mergedVolumeMounts, v)
+				continue
+			}
+
+			// make sure they are identical or throw an error
+			if !reflect.DeepEqual(found, v) {
+				errs = append(errs, fmt.Errorf("merging volume mounts for %s has a conflict on %s: \n%#v\ndoes not match\n%#v\n in container", pp.GetName(), v.Name, v, found))
+			}
 		}
 	}
 
-	// key by mount path
-	orig = map[string]interface{}{}
-	for _, v := range original {
-		orig[v.MountPath] = v
-	}
+	for _, pp := range podPresets {
```

Comment:
It seems like we can do all of this in a single nested loop, since this is orthogonal to the above loop but covers the same space